src/Confluent.Kafka/Consumer.cs

```diff
@@ -28,28 +30,160 @@
 namespace Confluent.Kafka
 {
     /// <summary>
-    ///     Implements a high-level Apache Kafka consumer (with 
-    ///     key and value deserialization).
+    ///     Implements a high-level Apache Kafka consumer.
     /// </summary>
-    public class Consumer<TKey, TValue> : IDisposable
+    public class Consumer<TKey, TValue> : IConsumer<TKey, TValue>
     {
-        private readonly Consumer consumer;
+        private bool disposeHasBeenCalled = false;
+        private object disposeHasBeenCalledLockObj = new object();
 
         /// <summary>
-        ///     The IDeserializer implementation instance used to deserialize keys.
+        ///     keeps track of whether or not assign has been called during
+        ///     invocation of a rebalance callback event.
         /// </summary>
-        public IDeserializer<TKey> KeyDeserializer { get; }
+        private bool assignCalled = false;
+
+        private readonly bool enableHeaderMarshaling = true;
+        private readonly bool enableTimestampMarshaling = true;
+        private readonly bool enableTopicNamesMarshaling = true;
+
+        private IDeserializer<TKey> KeyDeserializer { get; }
+        private IDeserializer<TValue> ValueDeserializer { get; }
+
+        private readonly SafeKafkaHandle kafkaHandle;
+
+        private static readonly byte[] EmptyBytes = new byte[0];
+
+
+        private Action<Error> errorDelegate = null;
+        private readonly Librdkafka.ErrorDelegate errorCallbackDelegate;
+        private void ErrorCallback(IntPtr rk, ErrorCode err, string reason, IntPtr opaque)
+        {
+            // Ensure registered handlers are never called as a side-effect of Dispose/Finalize (prevents deadlocks in common scenarios).
+            if (kafkaHandle.IsClosed) { return; }
+
+            if (errorDelegate != null)
```

Comment:
shouldn't there be a lock to protect this field?

Reply:
this can't change after construction, so it's all good.

Reply:
note: I have locking on the log callback because it can be called from any thread - the intent is to protect execution of the delegate (as opposed to the setting of it). this is possibly erring on the side of overly cautious, but I did it anyway.